src/server/decode.rs

```diff
@@ -116,13 +116,15 @@ fn set_url_and_port_from_host_header(req: &mut Request) -> http_types::Result<()
         .ok_or_else(|| format_err!("Mandatory Host header missing"))? //  https://tools.ietf.org/html/rfc7230#section-5.4
         .to_string();
 
-    if let Some(colon) = host.find(":") {
-        req.url_mut().set_host(Some(&host[0..colon]))?;
-        req.url_mut()
-            .set_port(host[colon + 1..].parse().ok())
-            .unwrap();
-    } else {
-        req.url_mut().set_host(Some(&host))?;
+    if !req.url().cannot_be_a_base() {
```

Comment:
It would be nice to have a test for this new condition, like “if the Url is cannot-be-a-base it does not attempt to set the host from Host header.” Did this panic previously?

Reply:
yeah, IIRC it would fail silently on CONNECT requests, since those headers are structured differently ie: `CONNECT www.example.com:443 HTTP/1.1`

Reply:
I'm not sure if my solution is the best one, since i think the problem might be with the url parser, but it fixed my issue... Feel free to modify as you see fit.

Reply:
I was just reading [the spec on connect](https://tools.ietf.org/html/rfc7231#section-4.3.6) and it seems like we should be able to set a host from that. Instead of changing this I think we probably want to change how we’re constructing the url before we hit this code, instead of leaving an unset host

Reply:
Ok, happy to park this PR until that is fixed then I can remove that part

Reply:
@Ben-Lichtman Now that #116 has been merged, the `req.url().host()` should be set correctly for CONNECT requests on master (and, importantly, not error)

Reply:
Ok, it should be good to go now